language_bindings/rust/rust_build/src/lib.rs

```diff
@@ -0,0 +1,360 @@
+///
+//TODO update rust docs
+use std::{env, io};
+
+/// This builder must have a path to a complied Halide folder. It is used to build multiple generators quickly
+///
+///
+pub struct GenBuilder {
+    halide_path: PathBuf,
+    gen_path: PathBuf,
+
+    rs_output: PathBuf,
+
+    //generators: Vec<Generator<'static>>,
+    debug: bool,
+    target: String,
+}
+
+/// This represents a Halide generator that is built using [GenBuilder]
+///
+///
+pub struct Generator<'a> {
+    gen_name: String,
+    gen_exe: PathBuf,
+    halide_path: PathBuf,
+    gen_path: PathBuf,
+
+    rs_output: PathBuf,
+
+    gcc_flags: Vec<&'a str>,
+    debug: bool,
+    target: String,
+}
+
+//Todo add runtime maker
+impl GenBuilder {
+    /// Instantiate a GenBuilder
+    ///
+    /// ```ignore
+    /// let Hal = GenBuilder::new(
+    ///       "Path/to/Halide",
+    ///        "Path/to/your/generators"
+    ///    )
+    /// ```
+    ///
+    pub fn new<T: Into<PathBuf>>(
+        halide_path: T,
+        gen_path: T,
+        //rs_output:T
+    ) -> GenBuilder {
+        GenBuilder {
+            halide_path: halide_path.into().join("distrib"),
+            gen_path: gen_path.into(),
+            rs_output: PathBuf::from(env::var("OUT_DIR").unwrap_or("target".to_string())),
+            //generators: vec![],
+            debug: false,
+            target: "target=host-no_runtime".to_string(),
+        }
+    }
+    /// Override the output directory.
+    ///
+    /// by default this is the ENV "out_dir" but can be anywere
+    /// useful to place .rs and headers in a folder in the src of your project but doing that
+    /// means cargo does not clean these artifacts
+    pub fn out_dir<T: Into<PathBuf>>(mut self, out: T) -> Self {
+        self.rs_output = out.into();
+        self
+    }
+    /*
+    pub fn push_gen(mut self, gen: Generator<'static>)->Self{
+        self.generators.push(gen);
+        self
+    }
+
+     */
+    //Todo template all strings
+    /// Create a generator based on its name
+    ///
+    /// We have strict naming conventions:   
+    ///    - The generator source must be gen_name.cpp   
+    ///    - The internal cpp class must also be gen_name
+    ///
+    ///```ignore
+    /// let gen = genBuilder.new_gen(iir_blur);
+    /// ```
+    pub fn new_gen(self, gen_name: String) -> Generator<'static> {
+        let mut temp_gen_exe = PathBuf::from(self.rs_output.to_str().unwrap());
+        if gen_name.is_empty() {
+            temp_gen_exe.push("runtime");
+        } else {
+            println!(
+                "cargo:rerun-if-changed={}",
+                self.gen_path
+                    .join(gen_name.as_str())
+                    .with_extension("cpp")
+                    .to_str()
+                    .unwrap()
+            );
+            temp_gen_exe.push(gen_name.as_str());
+        }
+        temp_gen_exe.with_extension("generator");
+
+        println!("temp: {:?}", temp_gen_exe.as_os_str());
```

Comment:
Avoid performing console I/O directly in library code, as it may interact poorly with downstream requirements. If you want to emit diagnostics, the `log` crate might be more appropriate, as it allows downstream code to filter and route diagnostics as desired. Alternatively, if this is debugging output for your own use and not of interest downstream at all, it should be removed.